### PR TITLE
feat: add SMI log collector

### PR DESCRIPTION
**Pull request by johnsonshi**
Add log collector for Service Mesh Interface
logs, CustomResourceDefinitions, and custom
resources.

Related to PR #48.

Resolves issue #67.

Signed-off-by: Johnson Shi <Johnson.Shi@microsoft.com>

### CODE REVIEW

**Comment by johnsonshi** [comment]
Hey @Tatsinnit, I created a draft PR outlining the types of logs of SMI CRDs that I'll be collecting. Let me know what you think plus your thoughts on how the logs are organized.

File: pkg/collector/smi_collector.go
Comment (johnsonshi) [inline pkg/collector/smi_collector.go:80]: @Tatsinnit (and @SanyaKochhar // @shalier) - thoughts on the types of SMI logs I'll collect plus the way I organized the files?

**Comment by johnsonshi** [comment]
Ready now

File: pkg/collector/smi_collector.go
Comment (johnsonshi) [inline pkg/collector/smi_collector.go:47]: Should this return error here?

File: deployment/aks-periscope.yaml
Comment (johnsonshi) [inline deployment/aks-periscope.yaml:83]: Will reverse this.

File: deployment/aks-periscope.yaml
Comment (johnsonshi) [inline deployment/aks-periscope.yaml:154]: Should I reverse this?

Comment (SanyaKochhar) [reply deployment/aks-periscope.yaml:154]: I think we had decided that by default these collectors will not be run, but can be enabled by users optionally. @Tatsinnit should we be adding documentation around this?

File: pkg/collector/smi_collector.go
Comment (SanyaKochhar) [inline pkg/collector/smi_collector.go:52]: nit: crdDefinitions = custom resource definition Definitions
```suggestion
	collectSmiCrds(collector, filepath.Join(rootPath, "smi_crd_definitions"), smiCrdsList)
```

Comment (johnsonshi) [reply pkg/collector/smi_collector.go:52]: thanks! :)

File: pkg/collector/smi_collector.go
Comment (Tatsinnit) [inline pkg/collector/smi_collector.go:63]: 💡 is this `fmt.Errorf(` or `log.Printf` I will let you decide, but the message `Failed` suggest its `Error`?

File: pkg/collector/smi_collector.go
Comment (Tatsinnit) [inline pkg/collector/smi_collector.go:72]: 💡 same as above: Worth identifying that use-case here is `error` or something else?

File: pkg/collector/smi_collector.go
Comment (Tatsinnit) [inline pkg/collector/smi_collector.go:77]: 💡Idea: Worth adding minor comment as to `namespace_*` leads to? Objective with the `join` aka `namespaceRootPath` docs from msdn or something?

File: pkg/collector/smi_collector.go
Comment (Tatsinnit) [inline pkg/collector/smi_collector.go:82]: 💡 The name: `collectSmiCustomResourcesFromSingleNamespace` suggestion to me that `SingleNamespace` is key but what is that `singlenamespace` please?

File: pkg/collector/smi_collector.go
Comment (Tatsinnit) [inline pkg/collector/smi_collector.go:87]: 💡 Same as above, decide between `error` or not.

File: pkg/collector/smi_collector.go
Comment (Tatsinnit) [inline pkg/collector/smi_collector.go:102]: 💡 `test func(string) bool` --  test? (shall we name this more meaningful and better)? 

* I think motivated from this code here: https://stackoverflow.com/questions/37562873/most-idiomatic-way-to-select-elements-from-an-array-in-golang

File: pkg/collector/smi_collector.go
Comment (Tatsinnit) [inline pkg/collector/smi_collector.go:96]: 💡 Let's think if this is `error` or not? (Messaging shows error and mostly currently the collector in error define it as error.

**Review by Tatsinnit — CHANGES_REQUESTED** [review]
💡 Thank you so much for this, I have added initial comments below. Really appreciate it.

File: deployment/aks-periscope.yaml
Comment (johnsonshi) [reply deployment/aks-periscope.yaml]: Resolved in d34fbb85991d6fa1822bcf29d4b0158c5b3af79d

File: deployment/aks-periscope.yaml
Comment (johnsonshi) [reply deployment/aks-periscope.yaml]: Resolved in d34fbb85991d6fa1822bcf29d4b0158c5b3af79d

Comment (johnsonshi) [reply deployment/aks-periscope.yaml]: Resolved in d34fbb85991d6fa1822bcf29d4b0158c5b3af79d

File: pkg/collector/smi_collector.go
Comment (johnsonshi) [reply pkg/collector/smi_collector.go]: Resolved in eeb2fdb

File: pkg/collector/smi_collector.go
Comment (johnsonshi) [reply pkg/collector/smi_collector.go]: Resolved in 5b56484 - this should be an error. If the user chooses to collect `SMI` data, then the scenario is that the user *has* `SMI` installed in their cluster, which is why they chose to collect `SMI` data. If they explicitly chose `SMI` data through the explicitly setting the flag, then this needs to return error if it fails to find any SMI CRDs in the cluster.

File: pkg/collector/smi_collector.go
Comment (johnsonshi) [reply pkg/collector/smi_collector.go]: This is resolved in 5b56484f608330da011cacf14314c75cf620ddad.
This should not return an error and should not give up trying to find the YAML definitions of the other SMI CRDs.
When periscope fails to obtain the YAML definition of a single SMI CRD, it should log a message noting that it skipped collecting the YAML of *only this* SMI CRD (should not error). Periscope should then proceed with collecting the YAML definitions of other SMI CRDs.

```go
	for _, smiCrd := range smiCrdsList {
		fileName := smiCrd + "_definition.yaml"
		kubeCmd := []string{"get", "crd", smiCrd, "-o", "yaml"}
		if err := collector.CollectKubectlOutputToCollectorFiles(rootPath, fileName, kubeCmd); err != nil {
			log.Printf("Skipping: unable to collect yaml definition of %s to %s: %+v", smiCrd, fileName, err)
                         ^^^ Do not give up and error - continue collecting the yaml definitions of the other SMI CRDs.
		}
	}
```

File: pkg/collector/smi_collector.go
Comment (johnsonshi) [reply pkg/collector/smi_collector.go]: Resolved in 5b56484.
This should return an error. Periscope needs to list all of the cluster's namespaces in order to proceed with collecting SMI custom resources. If Periscope fails to list the cluster's namespaces (`kubectl get namespaces...`), then something horrible has gone wrong in the cluster, so an error should be returned here.

```
	// Get all namespaces in the cluster
	namespacesList, err := utils.GetResourceList([]string{"get", "namespaces", "-o", "jsonpath={..metadata.name}"}, " ")
	if err != nil {
		return fmt.Errorf("Failed to collect SMI custom resources: unable to list namespaces in the cluster: %+v", err)
	}
```

File: pkg/collector/smi_collector.go
Comment (johnsonshi) [reply pkg/collector/smi_collector.go]: Resolved in 5b56484.
Periscope should not give up when it fails to list custom resources of a specific type.
Rather, it should log a message, skip listing custom resources of this type, and move on with listing custom resources of another type in the list.
```go
	for _, smiCrdType := range smiCrdsList {
		// get all custom resources of this smi crd type
		customResourcesList, err := utils.GetResourceList([]string{"get", smiCrdType, "-n", namespace, "-o", "jsonpath={..metadata.name}"}, " ")
		if err != nil {
			log.Printf("Skipping: unable to list custom resources of type %s in namespace %s: %+v", smiCrdType, namespace, err)
			continue
		}
```

File: pkg/collector/smi_collector.go
Comment (johnsonshi) [reply pkg/collector/smi_collector.go]: This is resolved in 5b56484.
If periscope is unable to collect the YAML definition of a single custom resource, it should not give up and `error`.
Instead, it should log a message and proceed with collecting the YAML definitions of other custom resources (of the same type).
```go
		customResourcesRootPath := filepath.Join(rootPath, smiCrdType+"_custom_resources")
		for _, customResourceName := range customResourcesList {
			fileName := smiCrdType + "_" + customResourceName + ".yaml"
			kubeCmd := []string{"get", smiCrdType, customResourceName, "-n", namespace, "-o", "yaml"}
			if err := collector.CollectKubectlOutputToCollectorFiles(customResourcesRootPath, fileName, kubeCmd); err != nil {
				log.Printf("Skipping: unable to collect yaml definition of %s (custom resource type: %s) to %s: %+v", customResourceName, smiCrdType, fileName, err)
			}
		}
```

File: pkg/collector/smi_collector.go
Comment (johnsonshi) [reply pkg/collector/smi_collector.go]: Resolved in 5075df7.
I renamed the function to `collectSmiCustomResourcesFromNamespace` because the previous name was misleading.
What `collectSmiCustomResourcesFromNamespace` does is collect all smi custom resources from a namespace. The newly-changed function name is more appropriate and less confusing.

File: pkg/collector/smi_collector.go
Comment (johnsonshi) [reply pkg/collector/smi_collector.go:74]: Resolved in e1fb508.
Periscope will iterate through each namespace in the cluster and attempt to collect SMI custom resources from each namespace. If SMI custom resources exist in a namespace, the YAML definitions of each custom resources will be stored in a sub-directory named `namespace_X`.

Here's how the directory structure of SMI collected logs look like.

image

image

image

image

File: pkg/collector/smi_collector.go
Comment (johnsonshi) [reply pkg/collector/smi_collector.go]: Resolved with ee5f839

File: deployment/aks-periscope.yaml
Comment (johnsonshi) [reply deployment/aks-periscope.yaml:31]: Hi @Tatsinnit, thanks for reviewing this, but I do not quite understand your comment? These permissions are required by the Periscope SMI collector in order to:
* obtain the YAML definition of SMI CRDs
* list SMI custom resources of each SMI CRD type
* obtain the YAML definition of each SMI custom resource.

The SMI spec dictates that each service mesh should implement SMI CRDs themselves while conforming to SMI standards. These (below) are examples of SMI CRDs that are installed as part of `Open Service Mesh`. Other service meshes such as `Linkerd` have their own implementation of SMI CRDs, which may be different.

```shell
» kubectl get crds | grep smi                                                                                                                                                    johnsonshi@DESKTOP-AQ60TOB
httproutegroups.specs.smi-spec.io            2021-06-22T21:32:33Z
tcproutes.specs.smi-spec.io                  2021-06-22T21:32:33Z
trafficsplits.split.smi-spec.io              2021-06-22T21:32:33Z
traffictargets.access.smi-spec.io            2021-06-22T21:32:33Z
udproutes.specs.smi-spec.io                  2021-06-22T21:32:33Z
```

Adding additional permissions to enable Periscope to obtain/list/get SMI CRDs shouldn't introduce any new intended behavior. The only positive effect is that this prevents Periscope from running into permissions during SMI log collection.

Regarding testing Periscope's permissions to list/obtain/get SMI CRDs, I think that would be a challenge. The SMI CRDs can only be installed into a cluster by *installing a full service mesh, such as OSM or Linkerd*. Periscope won't be able to run integration tests to test SMI CRD permissions without installing OSM/Linkerd.

Comment (Tatsinnit) [reply deployment/aks-periscope.yaml:31]: Hiya @johnsonshi not sure about above comment, but all I meant to mention is when you are testing this work with local published image? (Please do test the vanilla scenario? aka without SMI work as well) Thanks.

Comment (johnsonshi) [reply deployment/aks-periscope.yaml:31]: Oh sure thing! Yeah when I tested it with the locally published image (with SMI flag) it worked. I will try testing with the vanilla periscope image with plus set the corresponding flags and let you know the test results.

Comment (johnsonshi) [reply deployment/aks-periscope.yaml:31]: @Tatsinnit I've conducted the tests.

As can be seen here, the deployment yaml that I ran had the upstream vanilla aks-periscope image: https://paste.ubuntu.com/p/Mqt2q2v9Wc/

As can be seen here, the deployment yaml that I ran had the additional ClusterRole permissions for smi CRDs. https://paste.ubuntu.com/p/tjjMYPq56R/

This setup (vanilla upstream image WITH the additional SMI permissions in ClusterRole) successfully ran without error: https://paste.ubuntu.com/p/R34G4JF6vR/


----

I have successfully tested my changes with both locally published image and the vanilla upstream image.
Under both scenarios I tested _with_ and _without_ smi flag (`COLLECTOR_LIST: managedCluster OSM SMI # <custom flag>`).
Also under both scenarios I tested _with_ the additional SMI cluster role permissions.

I think this PR is now ready for merging. Let me know if you need anything else.

Comment (Tatsinnit) [reply deployment/aks-periscope.yaml:31]: Thank you for this, there is branch conflict also I believe ou will notice that there is a simplified collector implementation will help to simplify this more.

Thank you for taking care of SMI, and Thank-you to  @arnaud-tincelin for already sorting out the OSM work for this simplification, please let us know once conflict is resolved.

File: pkg/collector/smi_collector.go
Comment (arnaud-tincelin) [inline pkg/collector/smi_collector.go:48]: I believe this line produces a warning as there is nothing to format. I think you want to use error.New here

File: pkg/collector/smi_collector.go
Comment (arnaud-tincelin) [inline pkg/collector/smi_collector.go:59]: I think you need to rebase as this function does not exist anymore

File: pkg/collector/smi_collector.go
Comment (arnaud-tincelin) [inline pkg/collector/smi_collector.go:69]: This line produces a warning:
- no capital letter
- %w instead of %v
- as you are already in an error message you could simplify the text

`return fmt.Errorf("collect SMI custom resources: unable to list namespaces in the cluster: %w", err)`

**Review by arnaud-tincelin — CHANGES_REQUESTED** [review]

**Comment by johnsonshi** [comment]
Hi @Tatsinnit @arnaud-tincelin, I've resolved all conflicts and addressed @arnaud-tincelin's additional comments. I think this can now be merged - thank you!!! 😀🙏🙏

File: deployment/aks-periscope.yaml
Comment (Tatsinnit) [inline deployment/aks-periscope.yaml:122]: ⚠️ Be careful and please rotate this sasKey.

File: deployment/aks-periscope.yaml
Comment (Tatsinnit) [inline deployment/aks-periscope.yaml:83]: 💡 I think this needs to be reverted. Since this is your local image.

File: deployment/aks-periscope.yaml
Comment (Tatsinnit) [inline deployment/aks-periscope.yaml:121]: ⚠️ Due to security reasons I would recommend cycling this account name as well.

File: tests/aksperiscopeintegration_test.go
Comment (Tatsinnit) [inline tests/aksperiscopeintegration_test.go:80]: 💡 please add this into a new test file `helper_test.go` since this function is test of a function in `helper` and not really an integration test.

**Review by Tatsinnit — CHANGES_REQUESTED** [review]
👍 Rest looks good, once you made leftover recommended changes and especially not checking in with credentials :) 

cc: @arnaud-tincelin for rest. Thank you so much, 🙏

File: pkg/utils/helper.go
Comment (Tatsinnit) [inline pkg/utils/helper.go:367]: 💡 `ss` - is the acronym for?

Comment (johnsonshi) [reply pkg/utils/helper.go]: I have renamed this to `slice`

File: deployment/aks-periscope.yaml
Comment (johnsonshi) [reply deployment/aks-periscope.yaml]: Thanks for pointing this out! Removed and deleted.

**Review by Tatsinnit — APPROVED** [review]
👍 LGTM, thank you so much for this. 

I will defer to other maintainer folks for one last eyeballing, let’s merge it after that. Cc: @arnaud-tincelin since you’ve reviewed this before.  

Thanks 🙏

File: deployment/aks-periscope.yaml
Comment (arnaud-tincelin) [inline deployment/aks-periscope.yaml:31]: This file is deprecated and shall not be updated anymore. These changes should be added to the corresponding kustomize resource file instead.
@Tatsinnit do you agree?

**Review by arnaud-tincelin — CHANGES_REQUESTED** [review]

File: deployment/cluster-role.yaml
Comment (johnsonshi) [inline deployment/cluster-role.yaml:17]: Hi @arnaud-tincelin, I've updated the kustomize resource file so that CRDs and SMI CRDs (both required) are added to the periscope cluster role permissions.
`deployment/cluster-role.yaml`

File: deployment/aks-periscope.yaml
Comment (johnsonshi) [reply deployment/aks-periscope.yaml:31]: Hi @arnaud-tincelin, I think this should be resolved now that I've edited the corresponding kustomize resource file: https://github.com/Azure/aks-periscope/pull/71/files#r671532572

**Review by arnaud-tincelin — APPROVED** [review]

**Pull request merged**
